pkg/trace/metrics/timing/timing.go

```diff
@@ -0,0 +1,143 @@
+
+	"github.com/DataDog/datadog-agent/pkg/trace/atomic"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+)
+
+// AutoreportInterval specifies the interval at which the default set reports.
+const AutoreportInterval = 10 * time.Second
+
+var (
+	defaultSet = NewSet()
+	stopReport = defaultSet.Autoreport(AutoreportInterval)
+)
+
+// Since records the duration for the given metric name as time passed since start.
+// It uses the default set which is reported at 10 second intervals.
+func Since(name string, start time.Time) { defaultSet.Since(name, start) }
+
+// Flush reports the default set. It can be useful to call when the program
+// exits prematurely.
+func Flush() { stopReport() }
+
+// Set represents a set of metrics that can be used for timing. Use NewSet
+// to create one and Report (or Autoreport) to submit metrics. It is safe for
+// concurrent use.
+type Set struct {
+	mu sync.RWMutex        // guards c
+	c  map[string]*counter // maps names to their aggregates
+}
+
+// NewSet returns a new Set, optionally initialized with the given metric names.
+func NewSet(names ...string) *Set {
+	set := Set{
+		c: make(map[string]*counter, len(names)),
+	}
+	for _, name := range names {
+		set.c[name] = newCounter(name)
+	}
+	return &set
+}
+
+// Autoreport enables autoreporting of the Set at the given interval. It returns a
+// cancellation function.
+func (s *Set) Autoreport(interval time.Duration) (cancelFunc func()) {
+	stop := make(chan struct{})
+	go func() {
+		defer close(stop)
+		tick := time.NewTicker(interval)
+		defer tick.Stop()
+		for {
+			select {
+			case <-tick.C:
+				s.Report()
+			case <-stop:
+				s.Report()
+				return
+			}
+		}
+	}()
+	var once sync.Once // avoid panics
+	return func() {
+		once.Do(func() {
+			stop <- struct{}{}
+			<-stop
+		})
+	}
+}
+
+// Since records the duration for the given metric name as time passed since start.
+// If name does not exist, as defined by NewSet, it creates it.
+func (s *Set) Since(name string, start time.Time) {
+	s.mu.RLock()
+	c, ok := s.c[name]
+	s.mu.RUnlock()
+	if !ok {
+		c = newCounter(name)
+		s.mu.Lock()
+		s.c[name] = c
+		s.mu.Unlock()
+	}
+	ms := time.Since(start) / time.Millisecond
+	c.add(float64(ms))
+}
+
+// Report reports all of the Set's metrics to the statsd client.
+func (s *Set) Report() {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	for _, c := range s.c {
+		c.flush()
+	}
+}
+
+type counter struct {
+	// name specifies the name of this counter
+	name string
+
+	// mu keeps count and sum in sync, ensuring that average calculations
+	// are correct between add and flush calls.
+	mu    sync.RWMutex
+	sum   *atomic.Float64
+	count *atomic.Float64
+
+	max *atomic.Float64
+}
+
+func newCounter(name string) *counter {
+	return &counter{
+		name:  name,
+		count: atomic.NewFloat(0),
+		max:   atomic.NewFloat(0),
+		sum:   atomic.NewFloat(0),
+	}
+}
+
+func (c *counter) add(v float64) {
+	if v > c.max.Load() {
+		c.max.Store(v)
+	}
+	c.mu.RLock()
+	c.count.Add(1)
```

Comment:
Do we even need locking for these `atomic.Float64`? It looks like inside they do "compare and swap" until successful update, which means we don't need to lock here.

Reply:
That's not why the locking is there. I tried to explain it [here](https://github.com/DataDog/datadog-agent/pull/3423/files#diff-4bd5b7236c2211904bdb4b9dafd6f08bR106), in the documentation of `mu`. Basically I want to make sure that `count` and `sum` are always in sync, otherwise you might have issues when calculating the average (which is `sum/count`). For example, someone could be adding values while we are flushing, resulting in a discrepancy between `sum` and `count`, giving us the wrong average.

The lock here is a _read_ lock to allow multiple additions at once, while the flush lock is a _write_ lock to allow no modifications to `sum` and `count` while flushing. I'd say this is an unconventional use of locking. Does it make sense? Do you think there's a flaw in my logic?

Reply:
> I'd say this is an unconventional use of locking.

Oh, I see. That makes sense. Why not include `c.max` in the locking as well then? Then we could get a weird scenario where we flush out a "newer" `max` value along with an "older" `sum` and `count`?